lib/ansible/plugins/loader.py

```diff
@@ -1009,58 +1019,249 @@ class Jinja2Loader(PluginLoader):
-        raise AnsibleError('No code should call "get" for Jinja2Loaders (Not implemented) for non collection use')
+                for func_name, func in plugin_map:
+                    fq_name = '.'.join((parent_prefix, func_name))
+                    # TODO: load  anyways into CACHE so we only match each at end of loop
+                    #       the files themseves should already be cached by base class caching of modules(python)
+                    if key in (func_name, fq_name):
+                        pclass = self._load_jinja2_class()
+                        plugin = pclass(func)
+                        if plugin:
+                            context = plugin_impl.plugin_load_context
+                            self._update_object(plugin, fq_name, plugin_impl.object._original_path)
+                            break  # go to next file as it can override if dupe (dont break both loops)
+
+        except AnsiblePluginRemovedError as apre:
+            raise AnsibleError(to_native(apre), 0, orig_exc=apre)
+        except (AnsibleError, KeyError):
+            raise
+        except Exception as ex:
+            display.warning('An unexpected error occurred during Jinja2 plugin loading: {0}'.format(to_native(ex)))
+            display.vvv('Unexpected error during Jinja2 plugin loading: {0}'.format(format_exc()))
+            raise AnsibleError(to_native(ex), 0, orig_exc=ex)
+
+        return get_with_context_result(plugin, context)
 
     def all(self, *args, **kwargs):
+
+        # inputs, we ignore 'dedupe' we always do, used in base class to find files for this one
+        path_only = kwargs.pop('path_only', False)
+        class_only = kwargs.pop('class_only', False)    # basically ignored for test/filters since they are functions
+
+        # Having both path_only and class_only is a coding bug
+        if path_only and class_only:
+            raise AnsibleError('Do not set both path_only and class_only when calling PluginLoader.all()')
+
+        found = set()
+        # get plugins from files in configured paths (mulitple in each)
+        for p_map in self._j2_all_file_maps(*args, **kwargs):
+
+            # p_map is really object from file with class that holds mulitple plugins
+            plugins_list = getattr(p_map, self.method_map_name)
+            try:
+                plugins = plugins_list()
+            except Exception as e:
+                display.vvvv("Skipping %s plugins in '%s' as it seems to be invalid: %r" % (self.type, to_text(p_map._original_path), e))
+                continue
+
+            for plugin_name in plugins.keys():
+                if plugin_name in _PLUGIN_FILTERS[self.package]:
+                    display.debug("%s skipped due to a defined plugin filter" % plugin_name)
+                    continue
+
+                if plugin_name in found:
+                    display.debug("%s skipped as duplicate" % plugin_name)
+                    continue
```

Comment:
Yes, and this is in Jinja2Loader.all() so the 3 lines above would need to be removed in addition to the suggested diff.

Reply:
then you allow dupes in the normal loading for templates?

Reply:
normal loading for templates should be using fqcn though and it doesn't look like we dedupe regardless (on devel or with or without these lines). Maybe I'm misunderstanding, do you have a reproducer?

Reply:
we do 'both', first we load 'all' to get builtin/legacy and then if fqcn is present we get the specific one (used to force reload of all with new extra dirs, but this PR changes it to use loader.get functions).

Reply:
we 'could' make the dedupe=false parameter apply here and pass it in (it really only exists for base class to list the filter/test files since we dont dedupe on file name for those 2 plugins)

Reply:
It seems like there is a pre-existing bug (on devel and this branch) with using ansible.legacy filters that shadow a builtin name. The builtin is always used. ansible.builtin doesn't need this if ansible.legacy doesn't work because we know there aren't duplicates. Collections do have an issue with not deduping (last loaded wins), and this doesn't handle that.

Reply:
i've been trying to avoid adding the colleciton to .all() .. the other option i was looking at was adding a `list` method to the plugin loader that allows passing a collection name/path